04-onion-routing.md

```diff
@@ -1407,6 +1438,87 @@ The _origin node_:
   - MAY use the data specified in the various failure types for debugging
   purposes.
 
+## Recommendations for Reputation
+
+Local reputation is used by forwarding nodes to decide whether to endorse a HTLC
+on their outgoing link. Nodes MAY use any metric of their choosing to classify
+a peer as having "good" reputation, though a poor choice of reputation scoring
+metric may affect their downstream reputation.
+
+Peers build reputation by forwarding successful HTLCs that resolve quickly, and
+lose reputation if they endorse failing or slow-resolving HTLCs. Reputation is
+only _negatively_ affected if an endorsed HTLC resolves undesirably, to hold
+nodes accountable for their endorsement signal while still allowing them to 
+forward unendorsed HTLCs that they are not certain about.
+
+HTLC resolution time is assessed relative to a threshold that the node 
+considers to be a reasonable amount of time for a HTLC to resolve: 
+- `resolution_period`: the amount of time a HTLC is allowed to resolve in that 
+  is classified as "good" behavior (default: 10 seconds).
+- `resolution_time`: the time elapsed between `update_add_htlc` and its 
+  resolution (`update_fulfill_hltc` / `update_fail_hltc` / 
+  `update_fail_malformed_htlc`).
+
+Each HTLC's contribution to reputation is expressed by its `effective_fees` 
+which capture the fees that it paid and the opportunity cost that holding it 
+for `resolution_time` incurred. 
+- `fees`: the fees paid by a forwarded HTLC (as described in [BOLT #7](07-routing-gossip.md#htlc-fees), 
+  equal to 0 if the HTLC was not fulfilled).
+- `opportunity_cost`: `ceil ( (resolution_time - resolution_period) / resolution_period) * fees`
```

Comment:
For a worst-case `resolution_period` which is by far larger than `resolution_time` I can create a big enough **negative** `opportunity_cost` along a route.

Given that LN payments unfold over a big route, if I control the source & destination of a payment I can decide for how long to hold the payment for.

Given this layout
- `A -- B -- C -- D`
- I control `A` & `D`
- `B` and `C` are some very big routing nodes

We consider the past relationship of `B` and `C` to be very good (i.e when `B` forwards to `C` then they will be endorsed because they have accumulated a lot of `effective_fees` over the window of interest)

I can follow these steps:
- I make a few good payments from `A` to `D` until I notice that `endorsed` is turned on (effectively meaning that `B` now endorses me)
- Given access to the endorsed slots, I now create one or multiple payments from `A` to `D` that upon success would pay a big amount of `fees`.
- After I receive the `endorsed` HTLCs (that correspond to the above payments) on `D` I hold on to them for as long as possible by not releasing the preimage.
- Just before the CLTV timeout I fail the payment.
- `A -- B`, `B -- C`, `C -- D` all damage their local reputation by `fees - opportunity_cost`, where `opportunity_cost` can  be a big enough multiple of `fees`.


If the above flow is feasible, then a node that **just** earned the `endorsed` flag by their peer can now cause them reputation damage by far greater than what it cost them (in fees) to earn that flag.

Reply:
To mitigate this possible attack, perhaps `opportunity_cost` formula can have an upper limit? 

Something that can still allow it to create damage on `effective_fees` (in order to maintain the earn slow / lose fast attribute) but not large enough to be cause significant damage on other links further into the route, links with much stronger reputation relationships.